Flexit component fix for updated modbus

jorgror opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Fix bug in flexit-component that have caused flexit to not work since 2021.6
pyflexit first argument should be a ModbusSerialClient

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

jorgror committed:
pyflexit first argument should be a ModbusSerialClient
This component broke with 2021.6
I have tested this patch on my setup and it restores functionality

jorgror commented:
@janiversen I think you might be best suited to review this. I have very little experience with the home-assistant code base and has until now only been a user of home-assistant and not a contributor.

This was my take on solving a issue which broke my flexit climate. I'm not sure if this is a proper way of solving the issue, but it does work on my setup.

shaiu reviewed (commented):
How did the new HA version break this?

jorgror commented:
I don't have enough knowledge about the code-base to pinpoint exactly which commit/pr that introduced the error, but the rewrite of modbus component (mentioned in the changelog), which flexit component depends on has caused the issue. The problem is that we are passing a object of wrong type to pyflexit. It expects a ModbusSerialClient, but after 2021.6 it receives something else. From debugging on my local setup I found that the ModbusSerialClient was accessible in hub._client

janiversen commented:
Will take a look tomorrow

Rodahl88 commented:
Same problem here, no signal from CI66 Modbus Adapter.

2021-07-28 22:56:12 ERROR (MainThread) [homeassistant.components.websocket_api.http.connection] [140330271447504] 'ModbusHub' object has no attribute 'write_register'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 185, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1491, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1526, in _execute_service
    await handler.job.target(service_call)
  File "/usr/src/homeassistant/homeassistant/helpers/entity_component.py", line 213, in handle_service
    await self.hass.helpers.service.entity_service_call(
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 658, in entity_service_call
    future.result()  # pop exception if have
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 811, in async_request_call
    await coro
  File "/usr/src/homeassistant/homeassistant/helpers/service.py", line 695, in _handle_entity_call
    await result
  File "/usr/src/homeassistant/homeassistant/components/climate/__init__.py", line 455, in async_set_fan_mode
    await self.hass.async_add_executor_job(self.set_fan_mode, fan_mode)
  File "/usr/local/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/usr/src/homeassistant/homeassistant/components/flexit/climate.py", line 164, in set_fan_mode
    self.unit.set_fan_speed(self._fan_modes.index(fan_mode))
  File "/usr/local/lib/python3.9/site-packages/pyflexit/pyflexit.py", line 185, in set_fan_speed
    self._conn.write_register(
AttributeError: 'ModbusHub' object has no attribute 'write_register'

janiversen commented on homeassistant/components/flexit/climate.py:
```diff
@@ -64,7 +64,7 @@ def __init__(self, hub, modbus_slave, name):
         self._cooling = None
         self._alarm = False
-        self.unit = pyflexit(hub, modbus_slave)
+        self.unit = pyflexit(hub._client, modbus_slave)
```
Referencing the _client in the modbushub have a number of side-effects which I suppose is unwanted:
- _client is not guaranteed to point at a serial client, it can very well be None.
- you bypass all the error handling in the modbus integration, and would be better of dealing directly with pymodbus.

The real problem is that modbus is now async and thus the calls have changed. But since the call (I suppose one among a number) is hidden in the library pyflexit, I do not have suggestions on how to bypass that.

The owner of pyflexit needs to update the library, or even better inherit modbus climate in the flexit integration and add the flexit specialities.

Sorry I cannot be of much help here.

janiversen reviewed (changes requested):
See comment.

janiversen commented:
I took a look at the pyflexit library, I do not really understand why it exist, the code should have been integrated directly. I can see a major rewrite is needed. I could do it, but I have no way of testing the result so it does not make a lot of sense.

Rodahl88 commented:
Er ganske ny til Home assistant, finnes det noen omveier her?

Håvard

tor. 29. jul. 2021 kl. 15:29 skrev jan iversen ***@***.***>:

> ***@***.**** requested changes on this pull request.
>
> See comment.
> ------------------------------
>
> In homeassistant/components/flexit/climate.py
> <https://github.com/home-assistant/core/pull/53583#discussion_r679151047>:
>
> > @@ -64,7 +64,7 @@ def __init__(self, hub, modbus_slave, name):
>          self._heating = None
>          self._cooling = None
>          self._alarm = False
> -        self.unit = pyflexit(hub, modbus_slave)
> +        self.unit = pyflexit(hub._client, modbus_slave)
>
> Referencing the _client in the modbushub have a number of side-effects
> which I suppose is unwanted:
>
>    - _client is not guaranteed to point at a serial client, it can very
>    well be None.
>    - you bypass all the error handling in the modbus integration, and
>    would be better of dealing directly with pymodbus.
>
> The real problem is that modbus is now async and thus the calls have
> changed. But since the call (I suppose one among a number) is hidden in the
> library pyflexit, I do not have suggestions on how to bypass that.
>
> The owner of pyflexit needs to update the library, or even better inherit
> modbus climate in the flexit integration and add the flexit specialities.
>
> Sorry I cannot be of much help here.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/home-assistant/core/pull/53583#pullrequestreview-718093833>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIAHMEQGDSUHJ5LDJPRLFJLT2FJUJANCNFSM5BC7IZHA>
> .
>

jorgror commented:
I can give it a try rewriting this component to use the new async modbus component. Do you think I should base it on top of ModbusThermostat or implement it in a similar matter?

Anyways, this will be a huge challenge for me and will probably take some time. I have little experience in python and no experience with home-assistant core. But I guess this is a good place to start 😄

janiversen commented:
Time is not a problem, if you are up to the challenge, but in all fairness you are embarking on a bigger challenge.

If you decide to continue I can guide you a bit (if you want).

Another alternative is of course I do the programming and you do documentarion/testing.

jorgror committed:
Implemented async reading of modbus values
Stopped using pyflexit as this is outdated and not needed
Instead using async_pymodbus_call from ModbusHub class

jorgror committed:
Bugfix: Reading fan mode from wrong register

jorgror committed:
Implemented async writing
Set target temperature and fan mode using modbus call
Added some error handling

jorgror committed:
No longer require pyflexit

jorgror commented:
I have managed to get the component working based on async_pymodbus_call. I re-requested review to check if my implementation is a "proper" way to implement the modbus reading and writing. Comparing to other components I think there is a lot of further improvements to be added to Flexit.

janiversen commented on homeassistant/components/flexit/climate.py:
```diff
@@ -64,34 +75,65 @@ def __init__(self, hub, modbus_slave, name):
         return SUPPORT_FLAGS
 
-    def update(self):
+    async def async_update(self, now=None):
```
the now=None is used in modbus because we do our own scheduling, you do not need it.

janiversen commented on homeassistant/components/flexit/climate.py:
```diff
@@ -35,18 +44,20 @@
     """Representation of a Flexit AC unit."""
 
-    def __init__(self, hub, modbus_slave, name):
+    def __init__(self, hub: ModbusHub, modbus_slave, name):
```
When using types (which is a good thing) please use it for all parameters.

janiversen commented on homeassistant/components/flexit/climate.py:
```diff
@@ -64,34 +75,65 @@ def __init__(self, hub, modbus_slave, name):
+        self._current_temperature = await self._async_read_temp_from_register(
+            CALL_TYPE_REGISTER_INPUT, 9
+        )
+        self._current_fan_mode = self._fan_modes[
```
What happens if (for any reason) read return a bad value.

janiversen committed:
Review comments.

janiversen reviewed (changes requested):
Good job!! 

I found a number of small issues, only 1 functional, and I have added a commit that makes the changes, please have a look at it, dismiss it if you want to do things differently.

When you approve the extra commit or add the changes you want, I will approve and merge the PR. You might get other comments later from other reviewers.

janiversen commented:
I also rebased on the newest dev, just to secure there are no merge problems.

janiversen reviewed (approved):
LGTM.

jorgror commented:
Thanks for reviewing and improving the PR. I will try to further improve the quality of this component, but it will probably take some time.

On Sat, Aug 7, 2021 at 09:46, jan iversen ***@***.***> wrote:

> Merged [#53583](https://github.com/home-assistant/core/pull/53583) into dev.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, [view it on GitHub](https://github.com/home-assistant/core/pull/53583#event-5125509915), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AOP3F657RVUHRJAPY43RUPLT3TQDZANCNFSM5BC7IZHA).
> Triage notifications on the go with GitHub Mobile for [iOS](https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675) or [Android](https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email).

MartinHjelmare commented on homeassistant/components/flexit/climate.py:
```diff
@@ -35,18 +44,25 @@
     """Set up the Flexit Platform."""
     modbus_slave = config.get(CONF_SLAVE)
     name = config.get(CONF_NAME)
     hub = hass.data[MODBUS_DOMAIN][config.get(CONF_HUB)]
```
Side note: If integrations should be able to depend on the modbus integration and its hubs we should create a module level function in the modbus integration that returns the correct hub. Integrations are not allowed to directly access `hass.data` to get data from other integrations. The integration, modbus in this case, needs to expose an api for its resources.

janiversen replied:
I will make an extension to modbus exposing the hubs, and use flexit as example, except it later today or tomorrow.

However I think the long term solution is to add hass.get_domain, because there are really no need for integration to hold a list of the hub it "owns".

janiversen commented:
Added PR #54277 to resolve the late review sidetone.
